compiler/rustc_trait_selection/src/traits/select/mod.rs

```diff
@@ -1038,10 +1038,149 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
+                        debug!("CACHE MISS");
+                        self.insert_evaluation_cache(param_env, predicate, dep_node, result);
+                        previous_stack.cache.on_completion(dfn);
+                    }
+
+                    Ok(result)
+                }
             }
         })
     }
 
+    #[instrument(level = "debug", skip(self, previous_stack, obligation), ret)]
+    fn evaluate_uninhabited_predicate(
+        &mut self,
+        previous_stack: TraitObligationStackList<'_, 'tcx>,
+        obligation: &PredicateObligation<'tcx>,
+        ty: Ty<'tcx>,
+        module: Option<DefId>,
+    ) -> Result<EvaluationResult, OverflowError> {
+        let infcx = self.infcx;
+        let tcx = infcx.tcx;
+        match ty.kind() {
+            ty::Never => Ok(EvaluatedToOk),
+            ty::Tuple(tys) => {
+                self.evaluate_uninhabited_predicate_fields(previous_stack, obligation, module, tys.into_iter())
+            }
+            ty::Adt(adt, args) => {
+                // For now, unions are always considered inhabited
+                if adt.is_union() {
+                    return Ok(EvaluatedToErr);
+                }
+
+                // Non-exhaustive ADTs from other crates are always considered inhabited
+                if adt.is_variant_list_non_exhaustive() && !adt.did().is_local() {
+                    return Ok(EvaluatedToErr);
+                }
+
+                // The ADT is uninhabited if all its variants are uninhabited.
+                for variant in adt.variants() {
+                    if variant.is_field_list_non_exhaustive() && !variant.def_id.is_local() {
+                        // Non-exhaustive variants from other crates are always considered inhabited.
+                        return Ok(EvaluatedToErr);
+                    }
+                    // This variant is uninhabited is any of its fields is uninhabited.
+                    let field_tys = variant.fields.iter().filter_map(|field| {
+                        let ty = tcx.type_of(field.did).instantiate(tcx, args);
+                        match (field.vis, module) {
+                            // The field is public or we do not care for visibility.
+                            (ty::Visibility::Public, _) | (_, None) => Some(ty),
+                            (ty::Visibility::Restricted(from), Some(module)) => {
+                                if tcx.is_descendant_of(module, from) {
+                                    Some(ty)
+                                } else {
+                                    // From `module`, this field may be privately uninhabited.
+                                    // Do not consider it as uninhabited.
+                                    None
+                                }
+                            }
+                        }
+                    });
+                    match self.evaluate_uninhabited_predicate_fields(previous_stack, obligation, module, field_tys)? {
```

Comment:
We're not normalizing these field tys, are we? Does the old uninhabitedness check consider normalization?